# Getting Started In Colab

In [1]:
!pip install -q git+https://github.com/zhouy185/rosa-starter.git

In [2]:
import numpy as np
from starter import ZONES, TIME_BLOCKS, COSTS, PROMISE, delivery_times

In [3]:
times = delivery_times("Far West", "Fri/Sat eve", 45)
print(len(times), "orders; average delivery time", round(times.mean(), 1), 
"minutes")

210 orders; average delivery time 45.4 minutes


In [4]:
times

array([33.2, 51.3, 35.2, 32.6, 38.3, 41.2, 56.3, 33.8, 51.2, 49.4, 73. ,
       65.4, 38.6, 60.6, 40.4, 39.7, 76. , 44.3, 22.6, 53.4, 31.3, 34.8,
       63.5, 51.2, 41.9, 49.4, 54.4, 54.4, 57.4, 59. , 78.3, 59.7, 44.9,
       80.7, 37.1, 32.7, 62.1, 18. , 48.2, 56.2, 53.3, 30.2, 88.3, 25.8,
       36.2, 52.4, 34.4, 31.2, 49.3, 36.6, 50.4, 37.2, 47.1, 45.3, 86. ,
       39.5, 26.5, 35.8, 27.1, 29.4, 33.3, 37.5, 62.6, 37.6, 31.1, 52.4,
       51.4, 48.7, 45.6, 41.2, 46.4, 38.7, 55.2, 32.9, 61.6, 65.1, 22.5,
       32.1, 11.1, 38. , 43. , 63. , 50.8, 22.4, 33.3, 43.8, 35.4, 62.8,
       47.5, 29.4, 36.2, 46. , 48.4, 38. , 29.9, 29.2, 62.1, 51.2, 55.8,
       67.6, 40.4, 33. , 38. , 62.4, 52.9, 47.2, 49.9, 36.6, 57.1, 32.4,
       30.5, 42.5, 38.7, 34.1, 38.3, 26.7, 55.3, 33.5, 51.3, 32. , 37.5,
       60.9, 33.8, 45.9, 66.3, 53.2, 49.8, 54.3, 41.1, 26.6, 83.6, 39.1,
       36.2, 49.5, 60.7, 37.9, 48.1, 38.1, 29.5, 41.4, 38. , 32.9, 48. ,
       35.3, 31.1, 72.4, 56.3, 45.9, 78.1, 62. , 44

## Flag

- For the Streamlit app in Part III, add the line 
**git+https://github.com/zhouy185/rosa-starter.git to your app's
requirements.txt** so Streamlit Community Cloud can install it too.

# Part I. Describe: How bad is it, really?

#### A) For any provided promised delivery time and each pair of specified zone and time block, write a function that returns the percentage of late orders.

### Logic for Part 1
1. **Pseudocode** (**A:**): 
    - When we enter our variables, they need to be passed into the delivery_times function, which returns a array of values, we need to determine which values are over our threshold (45 minutes), in the case of the hardcoded promise. 
    - We can iterate over the array, using if/else logic to increment a (**lateness**) variable (a list that has +1 added for every value greater than promise)
    - We can ge the total count using len of the times varible.
    - We then return lateness over len times as the outout of the function to return the percentage of late deliveries per zone.
2. **Pseudocode** (Handling All inputs):
    - When we input the zone as all, of which there are three, we need delivery time to loop through zone one at a time going through value, collecting the total in addtion to all the late delivries while still using a fixed value for promist and timeblock.
    - *revised* Apply the same logic as above to timezone, when building out the function, using conditional logic start with the assumption that it's all by default, else use specififed value.
3. **Pseudocode** (**B:** Ranking the pairs):
    - To determine the rank of pairs in descending order of late arrival rate, we need to capture the output of each permutation (12) and then order them from greatest to least.
    - We could run each combination of LateOrders in a for loo, capturing the output (a combination of the zone,timeblock and promise) in a list and then sorting it. 

4. **Pseudocode** (**C:** returning the average delivery time):
    - We can reuse blocks of the code from LateOrders.
    - We can repurpose the logic that enables us to use (all) and a specific zone, along with the loop that iterates through each permutation. 
    - Moreover, from the sample code, we can use the time.mean method to capture the average when the input is a non-all value.
    - In order to handle the all:one, all:all, and one:all scenarios we need caputure a running total of the minutes across each combination and then divide by the total inputs (orders) 
5. **Pseudocode** (**D:** rank the pairs of zone and time block delivery time):
    - Going to repurpose code from part B, and C to return an enumerated list.
    - Using a for loop to iterate through each permutation, we'll store an averagedeliverytime tuple in the same ranked pairts list, sort and use the same for loop logic to print the list.
6. **Pseudocode** (**E:** Which measure, the late rate (b) or the average delivery time (d), is more relevant):
    - Fill


In [5]:
#Code for part A

def LateOrders(zone, time_block, promise=PROMISE, show_statement=True):
    selected_zones = ZONES if zone == "all" else [zone]
    selected_time_blocks = TIME_BLOCKS if time_block == "all" else [time_block]

    late_count = 0
    total_count = 0

    for selected_zone in selected_zones:
        for selected_time_block in selected_time_blocks:
            times = delivery_times(selected_zone, selected_time_block, promise)
            total_count += len(times)

            for duration in times:
                if duration > promise:
                    late_count += 1

    percentage = 100 * late_count / total_count
    if show_statement:
        zone_label = "all zones" if zone == "all" else zone
        time_block_label = "all time blocks" if time_block == "all" else time_block
        print(
            f"For zone {zone_label} and time block {time_block_label}, "
            f"with a promise of {promise} minutes, {percentage:.2f}% of orders "
            "will be late."
        )
    return percentage


# Specific zone and time block
LateOrders("Far West", "Fri/Sat eve")

# All zones, one time block
LateOrders("all", "Fri/Sat eve")

# One zone, all time blocks
LateOrders("Far West", "all")

# All zones and all time blocks
LateOrders("all", "all")

For zone Far West and time block Fri/Sat eve, with a promise of 45 minutes, 40.48% of orders will be late.
For zone all zones and time block Fri/Sat eve, with a promise of 45 minutes, 26.36% of orders will be late.
For zone Far West and time block all time blocks, with a promise of 45 minutes, 19.22% of orders will be late.
For zone all zones and time block all time blocks, with a promise of 45 minutes, 12.19% of orders will be late.


12.189054726368159

In [6]:
#Code for Part B
ranked_pairs = []

for zone in ZONES:
    for time_block in TIME_BLOCKS:
        late_rate = LateOrders(
            zone,
            time_block,
            promise=45,
            show_statement=False,
        )
        ranked_pairs.append((late_rate, zone, time_block))

ranked_pairs.sort(reverse=True)

for rank, (late_rate, zone, time_block) in enumerate(ranked_pairs, start=1):
    print(
        f"{rank}. {zone} / {time_block}: "
        f"{late_rate:.2f}% late"
    )

1. Far West / Fri/Sat eve: 42.38% late
2. North / Fri/Sat eve: 24.67% late
3. Central / Fri/Sat eve: 14.74% late
4. North / Weekday eve: 12.80% late
5. Far West / Weekday eve: 11.18% late
6. Far West / Other: 8.85% late
7. Far West / Lunch: 8.33% late
8. Central / Weekday eve: 6.36% late
9. North / Lunch: 5.26% late
10. North / Other: 5.00% late
11. Central / Lunch: 2.69% late
12. Central / Other: 1.58% late


In [7]:
#Code for part C
def AverageDeliveryTime (zone, time_block, promise=PROMISE):
    selected_zones = ZONES if zone == "all" else [zone]
    selected_time_blocks = TIME_BLOCKS if time_block == "all" else [time_block]

    TotalMinutes = 0
    TotalOrders = 0

    for selected_zone in selected_zones:
        for selected_time_block in selected_time_blocks:
            times = delivery_times(selected_zone, selected_time_block, promise)
            TotalMinutes += times.sum()
            TotalOrders +=len(times)
    return TotalMinutes / TotalOrders



In [8]:
print("Specific zone and specific time block:")
print(f"{AverageDeliveryTime('Far West', 'Fri/Sat eve', promise=PROMISE):.1f} minutes")

print("All zones and a specific time block:")
print(f"{AverageDeliveryTime('all', 'Fri/Sat eve', promise=PROMISE):.1f} minutes")

print("A specific zone and all time blocks:")
print(f"{AverageDeliveryTime('Far West', 'all', promise=PROMISE):.1f} minutes")

print("All zones and all time blocks:")
print(f"{AverageDeliveryTime('all', 'all', promise=PROMISE):.1f} minutes")

Specific zone and specific time block:
41.5 minutes
All zones and a specific time block:
38.6 minutes
A specific zone and all time blocks:
35.3 minutes
All zones and all time blocks:
32.3 minutes


In [9]:
#Code for Part D
RankedAveragePairs = []

for zone in ZONES:
    for time_block in TIME_BLOCKS:
        AverageDelivery = AverageDeliveryTime(
            zone,
            time_block,
            promise=PROMISE,
        )
        RankedAveragePairs.append((AverageDelivery, zone, time_block))

RankedAveragePairs.sort(reverse=True)

for rank, (AverageDelivery, zone, time_block) in enumerate(RankedAveragePairs, start=1):
    print(
        f"{rank}. {zone} / {time_block}: "
        f"has an Average Delivery time of {AverageDelivery:.1f} minutes "
    )

1. Far West / Fri/Sat eve: has an Average Delivery time of 42.8 minutes 
2. North / Fri/Sat eve: has an Average Delivery time of 38.4 minutes 
3. Far West / Weekday eve: has an Average Delivery time of 35.8 minutes 
4. Far West / Lunch: has an Average Delivery time of 33.1 minutes 
5. Central / Fri/Sat eve: has an Average Delivery time of 32.4 minutes 
6. Far West / Other: has an Average Delivery time of 31.8 minutes 
7. North / Weekday eve: has an Average Delivery time of 31.4 minutes 
8. Central / Weekday eve: has an Average Delivery time of 29.7 minutes 
9. North / Lunch: has an Average Delivery time of 28.9 minutes 
10. North / Other: has an Average Delivery time of 28.5 minutes 
11. Central / Lunch: has an Average Delivery time of 27.5 minutes 
12. Central / Other: has an Average Delivery time of 26.4 minutes 


**Part 1 E:**
- The late rate is more relevant to Rosa's decisions. Her promise is a threshold: every order that takes longer than 45 minutes is a broken promise, and a fast delivery to one customer doesn't make up for a slow one to another. Average delivery time blends these together and hides them. All 12 pairs average under 45 minutes, so based on the average alone, nothing looks wrong. Yet Far West / Fri/Sat eve averages 42.8 minutes and is late 46.2% of the time, because its delivery times are widely spread (some exceed 100 minutes).
- Average delivery time is still useful context, but the late rate points directly to where Rosa should act: either lengthen the promise for Far West on Friday and Saturday evenings, or add capacity there.

# Copilot support:

*prompt:* I'm providing the pseudocode / initial logic as captured in the markdown boxes, and asking copilot to review my approach and highlight issues or areas for improvemnt.

## Copilot (Part 1):
- Treat promise as a function input, using PROMISE as its default if appropriate, rather than hardcoding 45.
- Make lateness an integer counter, not a list.
- Multiply the fraction by 100 to return a percentage.
- Include both the zone and time block in the result’s scope.

**Implementing PsuedoCode for A**
- *Prompt:* I need help implementing this into code now. This is what've I tarted as a shell,but I'm getting tripped up with the conditional logic and the nesting of loops.
    - *Copilot (response) (Part A)*:
    - Copilot provided the first pass of the implementaion.

```python
      def LateOrders(zone, time_block, promise=PROMISE):
          selected_zones = ZONES if zone == "all" else [zone]
          selected_time_blocks = TIME_BLOCKS if time_block == "all" else [time_block]

          late_count = 0
          total_count = 0

          for selected_zone in selected_zones:
              for selected_time_block in selected_time_blocks:
                  times = delivery_times(selected_zone, selected_time_block, promise)
                  total_count += len(times)

                  for duration in times:
                      if duration > promise:
                          late_count += 1

          return 100 * late_count / total_count
```
- *Commentary:* From there, I had Copilot step me through the code, specifically the nested loops, the enforced Promise variable and then iterated to include printed output, indicating which zone, timeblock, and promise were used.

**Implementing PsuedoCode for B**

- *prompt:* 
    - Same structure as A 
- *Commentary:* 
    - I entered the pseudocode in the markdown above and walked through the output line by line with Copilot. 
    - The original implementation used a lambda function to achieve a more granular sort, which seemed out of scope for this course; moreover, it was superfluous in my view and not something I'm generally familiar with.

**Implementing PsuedoCode for C**

- *Commentary:* 
    - Given that this was mostly stitching together existing code, I implemnted the first iteration. 
    - However, the first iteration was overly complex and didn't account for all the scenarios as it performed the averaging too early in the process. 
```python
    def AverageDeliveryTime(zone, time_block, promise=PROMISE, show_statement=True):
    selected_zones = ZONES if zone == "all" else [zone]
    selected_time_blocks = TIME_BLOCKS if time_block == "all" else [time_block]

    
    AverageTime = 0

    for selected_zone in selected_zones:
        for selected_time_block in selected_time_blocks:
            times = delivery_times(selected_zone, selected_time_block, promise)
            AverageTime += round(times.mean(), 1)

    return AverageTime
```
*prompt:* 
   - Working on part C, I've stitched together a new function that returns the average delivery time. Highlight any issues with this logic, and walk me through the reasoning behind any issues you find before suggesting an alternative implementation.

**Implementing PsuedoCode for D**
- *Commentary:*
    - Again, this was hand-stitched; Copilot ran some validation and pointed out a few oversights on my part, but the implementation didn't really change.